element/daterange/settings.php

```diff
@@ -0,0 +1,40 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Admin settings.
+ *
+ * @package    customcertelement_daterange
+ * @copyright  2018 Dmitrii Metelkin <dmitriim@catalyst-au.net>
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+defined('MOODLE_INTERNAL') || die('Direct access to this script is forbidden.');
+
+$options = [];
+
+foreach (range(1, 365) as $number) {
+    $options[$number] = $number;
+}
+
+$settings->add(new admin_setting_configselect(
+    'customcertelement_daterange/maxranges',
```

Comment:
Why do we have this setting? Couldn't we just let the user add/remove them dynamically without a site setting being required?

Reply:
Or let the admin just add multiple elements?

Reply:
@markn86, as far as we know, there is no functionality and no real best practice in mform which would have supported us in building a feature to add dateranges dynamically. So we added an admin setting to define how many dateranges are available in a daterange element. 

In our local setup, this setting is set to 2 as we only want to support dateranges for winter term and summer term. Others may want to set it to 10 or something like that to support up to 10 dateranges.

If you have a good proposal how to add dateranges dynamically, I think I would not mind if you change this particular code.

Reply:
I think the setting needs more definition. I mean 365 date ranges? I doubt that would ever happen, and if it did it would be a messy looking certificate haha. Could you not just keep the element to one date range and if you needed two you could add two of the elements? Perhaps I am missing some functionality where adding two date ranges in one element does something that adding two elements would not.

Reply:
If you create a choice activity in a course, there is a button to add more options. I suspect this could be done here. Basically you would have a button to add a new date range, which would save the current element and then redirect to a page where they would be editing the newly created date range. It's not ideal as it creates the date range element before it has been finalised but I think relying on an admin setting is counter-intuitive. This would also mean having a button to remove a date range. Something similar is already done in the custom certificate when you 'add' a new page. What are your thoughts?

Reply:
Well, your hint with the choice activity was good. I saw that the button which you have referenced is built in https://github.com/moodle/moodle/blob/master/mod/choice/mod_form.php#L70-L71 and there seems to be a possibility to repeat elements in mforms which I simply wasn't aware of.

In choice activity, there isn't a button to remove choices and I think it would be tricky then to built it for daterange elements. However, in the choice activity, the form seems to be smart enough to only store these repeated form element groups which have been filled. If I create 5*3 form elements by clicking the button multiple times, but fill only four elements, 2*3 elements are presented when I open the form again for editing.

Summing up, we can try to have a look at changing the current admin setting to the "add element" button which you proposed, but there won't be a remove element button. Would this be also ok for you?

Reply:
Yep, that sounds fine Alex!

Reply:
@markn86 @abias  I have had a quick look at the code and it seems like repeat_elements method belongs to moodleform class, but we pass just $mform object to render_form_elements method so we can't in the current implementation handle repeated elements. 

@markn86 any suggestions about improving the plugin API to be able to create repeating elements? e.g add a second parameter to render_form_elements as pass edit_element_form instance?

Reply:
The $mform object is an instance of `edit_element_form` which extends `moodleform` so it should be available. I should use type hinting to make it obvious.

Reply:
@markn86 are you sure? It seems like $mform is MoodleQuickForm.

Reply:
`class edit_element_form extends \moodleform {` - is `moodleform` not what you are after?

Reply:
It does, but we pass just $mform to the element method. See https://github.com/markn86/moodle-mod_customcert/blob/MOODLE_36_STABLE/classes/edit_element_form.php#L67

Reply:
Ahh. My mistake. Grr. I should be passing `$this`. I would like to change this but am afraid of breaking third-party elements. How about copying the code and putting it in the element helper class? Not ideal, and it's a hack but think it's better than changing what we pass to `render_form_elements` because even if we make the second parameter optional PHP will complain (see https://3v4l.org/N0p4K).

Reply:
What about modifying get_element_instance method and set edit_element_form there?  We will introduce a new variable in the abstract class and will have set_element_form(edit_element_form $elementform) method to set it. Then in the factory we will set it after we initialise the element class.

Reply:
The approach sounds valid. The code should be pretty quick to write so i'll have a better idea if I could review it, but it sounds better than my suggestion of copy + pasting code that we will need to maintain.

Side note - looking at the code there are many things I want to re-write/refactor. The joys of software development.

Reply:
Just to clarify, is this the only thing that stops this code from being integrated?

Reply:
Yes.

I will integrate and then get a UX person to take a look and add issues later to improve it if there are any to do, but would be good to get the bulk of it in.

Reply:
Hi @markn86 Could you please have a look at the code I have just pushed? It's in this commit https://github.com/markn86/moodle-mod_customcert/pull/216/commits/87bda096a4b3ee2e05a9c70864dd58fbef28d0c5